firmware/examples/FactoryTest/source/main.cpp

```diff
@@ -0,0 +1,323 @@
+  //                   &control_sequence[0], 1);
+  // LOG_INFO("Control 2 = 0x%02X", control_sequence[0]);
+  // control_sequence[0] = 0x9F;
+  // i2c.Write(kGestureAddress, control_sequence, sizeof(control_sequence));
+
+  uint8_t initialization_sequence[] = { 0x80, 0x05 };
+  i2c.Write(kGestureAddress, initialization_sequence,
+            sizeof(initialization_sequence));
+
+  i2c.WriteThenRead(kGestureAddress, &kApds9960IdRegisterAddress, 1,
+                    &apds9960_id, 1);
+
+  LOG_INFO("Gesture ID = 0x%02X", apds9960_id);
+  LOG_INFO("End of Gesture Sensor Test...");
+
+  result = CheckDeviceId(apds9960_id, 0xAB);
+
+  return result;
+}
+
+bool AccelerometerTest()
+{
+  // Accelerometer Test
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting Accelerometer Test...\n\n");
+  uint8_t kMMAIdRegisterAddress     = 0x0D;
+  uint8_t mma_id                    = 0x00;
+  uint8_t initialization_sequence[] = { 0x2A, 0x01 };
+  bool result                       = false;
+
+  i2c.Write(kAccelerometerAddress, initialization_sequence,
+            sizeof(initialization_sequence));
+
+  i2c.WriteThenRead(kAccelerometerAddress, &kMMAIdRegisterAddress, 1, &mma_id,
+                    1);
+
+  LOG_INFO("Accelerometer ID = 0x%02X", mma_id);
+  LOG_INFO("End of Accelerometer Test...");
+
+  result = CheckDeviceId(mma_id, 0x2A);
+
+  return result;
+}
+
+bool TemperatureSensorTest()
+{
+  // Temperature Test
+  // ID should be 0x01 [7:4]
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting Temperature Test...\n\n");
+  uint8_t kTemperatureIdAddress = 0xC0;
+  uint8_t temperature_id        = 0x00;
+  bool result                   = false;
+
+  i2c.WriteThenRead(kTemperatureAddress, &kTemperatureIdAddress, 1,
+                    &temperature_id, 1);
+  LOG_INFO("ID:%02X", temperature_id);
+  result = CheckDeviceId(temperature_id, 0x14);
+
+  uint8_t control_sequence[] = { 0xC4, 0x00, 0xC9, 0x01 };
+  i2c.Write(kTemperatureAddress, control_sequence, sizeof(control_sequence));
+
+  // MSB, LSB for temperature data
+  uint8_t temperature_ms = 0xC1;
+  uint8_t temperature_ls = 0xC2;
+  int temperature_data   = 0;
+  i2c.WriteThenRead(kTemperatureAddress, &temperature_ms, 1, &temperature_ms,
+                    1);
+  i2c.WriteThenRead(kTemperatureAddress, &temperature_ls, 1, &temperature_ls,
+                    1);
+
+  temperature_data = ((temperature_ms & 0x7F) << 8) | temperature_ls;
+  LOG_INFO("temperature_data is %02X%02X, %04X", temperature_ms, temperature_ls,
+           temperature_data);
+  // Compute the actual temperature in Celsius
+  int actual_temp = 55 + ((temperature_data - 16384) / 160);
+  LOG_INFO("Temperature is: %u", actual_temp);
+  LOG_INFO("End of temperature test...");
+  if (-40 < actual_temp && actual_temp < 125)
+  {
+    result &= true;
+  }
+
+  return result;
+}
+
+bool CheckDeviceId(uint8_t actual_id, uint8_t expected_id)
```

Comment:
This function is pretty useless. It would actually be more useful to simply just do the equation in the code where this is called. What this should do is the following:

```
bool CheckDeviceId(I2c & i2c_peripheral, uint8_t device_address, uint8_t id_register_address, uint8_t  expected_id)
{
   // Do I2c transaction to figure stuff out here
}
```

If you notice, all of your id check calls have the same pattern, with a few values made differently. So you can make a function that actually does the i2c transaction for you and simply pass it a few bits of info about the device.